framework/include/utils/PetscVectorReader.h

```diff
@@ -0,0 +1,62 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "libmesh/petsc_vector.h"
+#include "libmesh/id_types.h"
+#include "libmesh/libmesh_common.h"
+#include "MooseError.h"
+
+using libMesh::Number;
+using libMesh::numeric_index_type;
+using libMesh::PetscVector;
+
+/**
+ * A class which helps with repeated reading from a ghosted petsc vector.
+ * Its main purpose is to avoid unnecessary calls to the get_array() function
+ * in the wrapper.
+ */
+class PetscVectorReader
+{
+public:
+  /// Construct using a pets vector
+  PetscVectorReader(PetscVector<Number> & vec);
+
+  /// Construct using a numeric vector
+  PetscVectorReader(NumericVector<Number> & vec);
+
+  /// Destructor to make sure the vector is restored every time this
+  /// goes out of scope
+  ~PetscVectorReader();
+
+  /// Restore the array, usually upon going out of scope
+  void restore();
+
+  /// Check if this vector is readable
+  bool readable() const { return _raw_value != nullptr; }
```

Comment:
Let's either make these private or just move their code into the associated public methods. Then we can be confident this is only used in RAII contexts